### PR TITLE
Adding some small missing commands in final Dockerfile

### PR DESCRIPTION
**Pull request by dawidmalina**
Adding some small missing commands in final Dockerfile:
- ability to copy whole resource directory in one command
- ability to add run commands (multiple run commands will be squash to one layer)
- ability to provide user
- convert windows paths (backslash) to unix path separator
- propopse new commands order: FROM, MAINTAINER, ENV, WORKDIR, ADD, RUN, EXPOSE, USER, ENTRYPOINT, CMD

### CODE REVIEW

**Comment by garethahealy** [comment]
+1 - be nice if this was merged.

**Comment by jleskovar** [comment]
Seconding. This seems really useful, and would save having to create an external Dockerfile in a lot of cases. It also enables a workaround for #58

**Comment by dawidmalina** [comment]
@davidxia can you tell us when this change can be merged? Looks like so many people are waiting for this PR to merged.

File: src/main/java/com/spotify/docker/BuildMojo.java [hunk 1]
```diff
@@ -118,6 +119,22 @@
...
+  private String workdir;
+
+  /** The user for the image. Ignored if dockerDirectory is set */
+  @Parameter(property = "user")
+  private String user;
+
+  /** 
+   * The run commands for the image. 
+   *  */
+  @Parameter(property = "dockerRuns")
```
Comment (gilles) [inline src/main/java/com/spotify/docker/BuildMojo.java]: Awesome! I was going to add that. One question I have: is it guaranteed that the order in the `pom.xml` file will be the order received here?

If not it might be a problem

Comment (dawidmalina) [inline src/main/java/com/spotify/docker/BuildMojo.java]: We did not used any special method to guarantee that. But we have never faced any issues with that. Run order from pom.xml was always kept unchanged.

Comment (gilles) [inline src/main/java/com/spotify/docker/BuildMojo.java]: sweet

**Comment by davidxia** [comment]
@dawidmalina Hey sorry I've been pretty busy. Thanks for this contribution, we'll take a look at this soon.

**Comment by davidxia** [comment]
@dawidmalina Do you want to rebase on top of master and tidy up the commit history. It's a little hard to read right now with the 33 commits. I'll take a stab at this if you don't get to this in time.

File: src/main/java/com/spotify/docker/BuildMojo.java [hunk 2]
```diff
@@ -404,11 +422,22 @@ private void loadProfile() throws MojoExecutionException {
       exposesSet.add(expand(raw));
     }
 
+    List<String> runListTmp = emptyList();
+    try {
+        runListTmp = profileConfig.getStringList("runs");
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/BuildMojo.java]: Use 2 space indents.

File: src/main/java/com/spotify/docker/BuildMojo.java [hunk 3]
```diff
@@ -486,42 +524,6 @@ private void createDockerFile(String directory, List<String> filesToAdd) throws
     if (maintainer != null) {
       commands.add("MAINTAINER " + maintainer);
     }
-    if (entryPoint != null) {
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/BuildMojo.java]: Why is this moved after env?

File: src/main/java/com/spotify/docker/BuildMojo.java [hunk 4]
```diff
@@ -565,19 +617,31 @@ private void createDockerFile(String directory, List<String> filesToAdd) throws
...
         getLog().info(String.format("Copying %s -> %s", sourcePath, destPath));
         // ensure all directories exist because copy operation will fail if they don't
         Files.createDirectories(destPath.getParent());
         Files.copy(sourcePath, destPath, StandardCopyOption.REPLACE_EXISTING,
                    StandardCopyOption.COPY_ATTRIBUTES);
-        // file location relative to docker directory, used later to generate Dockerfile
-        final Path relativePath = Paths.get(targetPath, included);
-        copiedPaths.add(relativePath.toString());
+        
+        if (!copyWholeDir) {
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/BuildMojo.java]: Shouldn't we check `copyWholeDir` outside of this for loop to be faster?

```java
if (copyWholeDir) {
  final Path relativePath = Paths.get(targetPath);
  copiedPaths.add(unixPath(relativePath));
} else {
  // above lines
}

File: src/main/java/com/spotify/docker/BuildMojo.java [hunk 5]
```diff
@@ -593,4 +657,28 @@ private void createDockerFile(String directory, List<String> filesToAdd) throws
...
+
+    /**
+     * When running on windows, paths are backslash separated,
+     * we need to convert them to unix '/' separated paths.
+     * 
+     * @param path
+     *            object (system specific)
+     * @return unix style path
+     */
+    private String unixPath(Path path) {
```
Comment (davidxia) [inline src/main/java/com/spotify/docker/BuildMojo.java]: Thanks. This should help all our neglected windows users :)

**Comment by davidxia** [comment]
I've created https://github.com/spotify/docker-maven-plugin/pull/121

**Comment by dawidmalina** [comment]
Tomorrow I will provide changes to cover all yours suggestions.

**Comment by davidxia** [comment]
@dawidmawlina actually can you take a look at https://github.com/spotify/docker-maven-plugin/pull/121 once I've cleaned up the git history? I've made the changes there.

**Comment by davidxia** [comment]
@dawidmalina OK, I've rebased https://github.com/spotify/docker-maven-plugin/pull/121. Please take a look. Thanks
